### PR TITLE
Validate HTTP status of data source URL

### PR DESCRIPTION
**Pull request by sbeh**
This should fix issues #117 and #166 .

A HTTP HEAD request is started per cities/city.js file and outputs a warning for non-200 response codes.

Those requests are done asynchronously, so I had to change the workflow of the validator script a bit:
- Setting exit code of the validating process has to wait until all requests are finished.
- Errors and warnings from async tasks are collected in lists.
- After all tasks have been finished, the lists are outputted and the exit code gets set.

### CODE REVIEW

**Comment by sbeh** [comment]
Shortened some lines of code.

File: validation/markets-validator.js [hunk 1]
```diff
@@ -464,7 +481,48 @@ function MetadataValidator(metadata) {
...
+        request = request.request(url, this.validateUrlStatusCallback);
+        request.on('error', function (error) {
+            asyncWarnings.push(new CustomIssue(
+                cityName + ': Error while accessing data source url: ' + error
+            ));
+        });
+        request.end();
+    };
+
+    this.validateUrlStatusCallback = function (response) {
```
Comment (johnjohndoe) [inline validation/markets-validator.js]: - Please remove space here `function (response)` for consistency.
- Please use `{ ... }` after an `if` statement.
- Please use a dedicated _issue_ such as:

``` js
function HttpResponseStatusIssue(cityName, statusCode) {

    this.cityName = cityName;
    this.statusCode = statusCode;

    this.toString = function() {
        return this.cityName + ": HTTP response status of data source url was: " + statusCode;
    };
}
```

File: validation/markets-validator.js [hunk 2]
```diff
@@ -464,7 +481,48 @@ function MetadataValidator(metadata) {
             this.errors.push(new NullAttributeIssue("url"));
         } else if (url.length === 0) {
             this.errors.push(new EmptyAttributeIssue("url"));
-        }
+        } else {
+            this.validateUrlStatus(url);
+        }
+    };
+
+    this.validateUrlStatus = function (url) {
```
Comment (johnjohndoe) [inline validation/markets-validator.js]: - Please remove space here `function (url)` for consistency.
- Please use a dedicated _issue_ such as:

``` js
function UnknownProtocolIssue(cityName, protocol) {

    this.cityName = cityName;
    this.protocol = protocol;

    this.toString = function() {
        return this.cityName + "Unknown protocol '" + this.protocol + "' in data source url.";
    };
}
```

**Comment by torfsen** [comment]
The problem with breaking the build on any faulty URL is that URLs usually go bad after they've been correct. In these cases we would end up with a now-faulty URL contributed and maintained by a non-core developer which breaks the complete build.

Faulty URLs are bad, but in my opinion not bad enough to break the build. I'd prefer a warning instead.

**Comment by johnjohndoe** [comment]
I fear nobody takes care if warnings are appended to the end of the build log. Could we at least notify someone?

**Comment by sbeh** [comment]
I like the idea of making all HTTP requests synchronous, I actually don't know how to do it ;)

**Comment by johnjohndoe** [comment]
Please note that we are fine with edits on **your branch** which rewrite the history. This means we prefer **clean atomic commits**. So if you **integrate review comments** please apply the changes to the respective commits. At the end you are welcome to **update** your branch and this pull request via `git push --force origin checkDataSourceUrl`.

**Comment by johnjohndoe** [comment]
Thank you for the edits.
Still we have to discuss how to avoid that someone has to manually inform the market data contributors from cities where URLs are outdated.

**Comment by sbeh** [comment]
It would be nice if you could `git blame -L 13,13` the contributor and somehow message him, but I don't think github allows this.

**Comment by sbeh** [comment]
Please keep in mind, currently this patch has another flaw:
It won't give you the cityName for bad urls, because I'm still trying to make it synchronous which then would print the error/warning right below the city.js filename.

**Comment by torfsen** [comment]
If we had a full build system (similar to the one I prototyped in #126) then we could validate cities separately and only include those without errors. But even then we would risk losing a city in the UI just because its URL has gone bad.

@k-nut Can we send warnings from the validation from Travis to Slack?

It seems that the [default Travis notifications](https://docs.travis-ci.com/user/notifications) don't allow you to customize the content. I was thinking about just sending the warnings after a successful run.

**Comment by k-nut** [comment]
@torfsen yes we can. Since we are just executing arbitrary code on Travis anyway we can add some kind of library to the validation script that notifies slack. This means that we would not be utilizing the built in Travis notifications but would have to write custom code but I do not think that it is much effort.

**Comment by torfsen** [comment]
@k-nut While that's not really what I had originally in mind  (using some kind of Travis built-in feature) it is probably a good idea! If there are warnings let the validation script check if we're running on Travis and in that case send a notification.

**Comment by sbeh** [comment]
@johnjohndoe I couldn't find a nice way to make those HTTP requests synchronous, so I prefixed all async warnings with the cityNames again.

**Comment by johnjohndoe** [comment]
@sbeh Maybe wait until we together discussed the approach to inform contributors about invalid data source URLs.
@k-nut I would prefer sending a notification directly to the contributor(s) of the city. A Slack notification puts the burden on us again to manually forward the task to them. Do you think this is possible?

**Comment by torfsen** [comment]
@johnjohndoe I'm not sure whether that's a good idea: Do you simply want to inform the person who contributed the PR? They might have worked on something completely unrelated to the URL.

Having a separate maintenance contact for each city would be possible but would probably be a lot more work to organize (and keep current) then just fixing the URL.

**Comment by johnjohndoe** [comment]
@torfsen I imagine of having (a) person(s) for each city feeling responsible to check the up-to-dateness of the market data of her city from time to time. This mixes in with a contact for non-technical website users mentioned in #86.

The migration of adding such a mandatory _contact_ field to the _<cityname>.json_ files could happen fluently. I would introduce a schema version for _<cityname>.json_ files. But let's put this discussion in a separate thread if needed.

**Comment by k-nut** [comment]
@johnjohndoe sure it would be possible to also contact individual people. Since we can execute any code we like only the sky is the limit ;)

**Comment by johnjohndoe** [comment]
I tend to merge the current state since it is already an enhancement of the current validation. We can discuss details in a follow-up issue and improve gradually. What do you think, people?
@sbeh I will review your code once again.

**Comment by torfsen** [comment]
@johnjohndoe While I would love to have a maintainer for each city I do think that managing them would take more effort than we would gain from having them in the first place.

But I agree that we can review and merge this PR as it is and then discuss that matter later on.

**Comment by johnjohndoe** [comment]
@sbeh Thank you very much for this contribution! :ok_hand: I rebased and finally merged your branch here: c50041df56a2944c89d52f56b09739a75c52d4f3 :santa:

**Comment by johnjohndoe** [comment]
Hm.. Somehow, the result of the async tasks does not show up in Travis CI: https://travis-ci.org/wo-ist-markt/wo-ist-markt.github.io/builds/196708860

**Pull request closed without merging**